pkg/test/kubeharness.go

```diff
@@ -0,0 +1,272 @@
+	}
+
+	if h.client == nil {
+		client, err := client.New(h.restConfig, client.Options{})
+		if err != nil {
+			h.Fatalf("error building client: %v", err)
+		}
+		h.client = client
+	}
+
+	logging.SetupLogger()
+
+	if loadCRDs {
+		crds, err := crdloader.LoadAllCRDs()
+		if err != nil {
+			h.Fatalf("error loading crds: %v", err)
+		}
+		{
+			var wg sync.WaitGroup
+			var errsMutex sync.Mutex
+			var errs []error
+
+			for i := range crds {
+				crd := &crds[i]
+				wg.Add(1)
+				log.V(2).Info("loading crd", "name", crd.GetName())
+
+				go func() {
+					defer wg.Done()
+					if err := h.client.Create(ctx, crd.DeepCopy()); err != nil {
+						errsMutex.Lock()
+						defer errsMutex.Unlock()
+						errs = append(errs, fmt.Errorf("error creating crd %v: %w", crd.GroupVersionKind(), err))
+					}
+					h.waitForCRDReady(crd)
+				}()
+			}
+			wg.Wait()
+			if len(errs) != 0 {
+				h.Fatalf("error creating crds: %v", errors.Join(errs...))
+			}
+		}
+	}
+
+	return h
+}
+
+func (h *KubeHarness) GetClient() client.Client {
+	return h.client
+}
+
+func (h *KubeHarness) GetRESTConfig() *rest.Config {
+	return h.restConfig
+}
+
+func (h *KubeHarness) waitForCRDReady(obj client.Object) {
+	logger := log.FromContext(h.Ctx)
+
+	apiVersion, kind := obj.GetObjectKind().GroupVersionKind().ToAPIVersionAndKind()
+	name := obj.GetName()
+	namespace := obj.GetNamespace()
+
+	id := types.NamespacedName{Name: name, Namespace: namespace}
+	if err := wait.PollImmediate(2*time.Second, 1*time.Minute, func() (bool, error) {
+		u := &unstructured.Unstructured{}
+		u.SetAPIVersion(apiVersion)
+		u.SetKind(kind)
+		logger.V(2).Info("Testing to see if resource is ready", "kind", kind, "id", id)
+		if err := h.GetClient().Get(h.Ctx, id, u); err != nil {
+			logger.Info("Error getting resource", "kind", kind, "id", id, "error", err)
+			return false, err
+		}
+		objectStatus := dynamic.GetObjectStatus(h.T, u)
+		// CRDs do not have observedGeneration
+		for _, condition := range objectStatus.Conditions {
+			if condition.Type == "Established" && condition.Status == "True" {
+				logger.V(2).Info("crd is ready", "kind", kind, "id", id)
+				return true, nil
+			}
+		}
+		// This resource is not completely ready. Let's keep polling.
+		logger.V(2).Info("CRD is not ready", "kind", kind, "id", id, "conditions", objectStatus.Conditions)
+		return false, nil
+	}); err != nil {
+		h.Errorf("error while polling for ready on %v %v: %v", kind, id, err)
+		return
+	}
+}
+
+func (h *KubeHarness) MustReadFile(p string) []byte {
+	return MustReadFile(h.T, p)
+}
+
+func (h *KubeHarness) EnsureNamespaceExists(name string) {
+	ctx := h.Ctx
+	ns := &corev1.Namespace{}
+	ns.SetName(name)
+	if err := h.GetClient().Create(ctx, ns); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			h.Fatalf("error creating namespace %v: %v", name, err)
+		}
+	}
+}
+
+// CreateDummyCRD registers a CRD so we can create objects in tests
+func (h *KubeHarness) CreateDummyCRD(group, version, kind string) {
```

Comment:
This is cool. (What do we do today?  I guess we should have such a CreateCRD already, but cannot find it via code search)

Reply:
I found it ... there's a function called `test.FakeCRDs()` that is called via a convoluted call chain.

I'll harmonize this implement in a follow-on PR.